gridmet_aggregation_PRMS.py

```diff
@@ -0,0 +1,105 @@
+        gdf = shp
+    elif os.path.isfile(shp):
+        gdf_prms_path_edited = shp
+        gdf = gpd.read_file(gdf_prms_path_edited, layer=gpkg_layer)
+    else:
+        print('shp must be path to geospatial file or a geodataframe')
+
+    ## Merge ncdf w/ shapefile (the shpfile has area info) & convert to GeoDataFrame
+    gridmet_drb_df = xr_mapped_df.merge(gdf, how ='left', left_on = left_on, right_index = right_on_index)
+    gridmet_drb_gdf = gpd.GeoDataFrame(gridmet_drb_df)
+
+    return gridmet_drb_gdf
+
+# gridmet_prms_area_avg_agg() function aggregates to groupby_cols scale and outputs the aggregated dataset
+def gridmet_prms_area_avg_agg(df, groupby_cols, val_colnames, wgt_col, output_path = None):
+    """
+    :param pd.DataFrame or gpd.GeoDataFrame df:
+    :param  list groupby_cols: group by cols of df
+    :param list val_colnames: list of columns to perform area-weighted average calculation on
+    :param str wgt_col: weight column
+    :param str output_path: if not None, path to save output df dataset in local dir.
+    :return: df with gridmet metrics per PRMS_segid
+    """
+
+    ## Weighted average calc = sum((val[1] * weight[1]), ... , (val[n] * weight[n]) ) / sum(weight[1],...,weight[n])
+
+    start = time.perf_counter()
+
+    df = df.copy()
+    ## 1. Calc value x weight col
+    for col in val_colnames:
+           df[f'{col}_x_area'] = df[col]*df[wgt_col]
+
+    ## 2. Create group by, by summing new value cols weighted by area
+    df_grouped = df.groupby(groupby_cols).agg(area_m2_sum = (wgt_col, 'sum'),
+                                              tmmx_wgtd_sum = ('tmmx_x_area', 'sum'),
+                                              tmmn_wgtd_sum = ('tmmn_x_area', 'sum'),
+                                              pr_wgtd_sum = ('pr_x_area', 'sum'),
+                                              srad_wgtd_sum = ('srad_x_area', 'sum'),
+                                              vs_wgtd_sum = ('vs_x_area', 'sum'),
+                                              rmax_wgtd_sum = ('rmax_x_area', 'sum'),
+                                              rmin_wgtd_sum = ('rmin_x_area', 'sum'),
+                                              sph_wgtd_sum = ('sph_x_area', 'sum')
+                                              )
+    ## 3. Get weighted average
+    for col in val_colnames:
+        df_grouped[f'{col}_wgtd_avg'] = df_grouped[f'{col}_wgtd_sum'] / df_grouped['area_m2_sum']
+
+    # remove interim col
+    df_final = df_grouped.loc[:, ~(df_grouped.columns.str.contains('wgtd_sum'))]
+
+    if output_path is not None:
+        df_final.to_csv(output_path, sep = ',')
+        print('Output saved in: ' + output_path)
+
+    end = time.perf_counter()
+
+    print('Time (sec) ellapsed:', end - start)
+
+    return df_final
+
+# Define variables and run
+if __name__ =='__main__':
+
+    ## Variable definitions
+    gdf_prms_path_edited = './data/GFv1_catchments_edited.gpkg'
```

Comment:
Instead of committing this file, what do you think of pulling it from the `drb-network-prep` repo like we have done in IS repo (i.e., with this URL: GFv1_HRUs_edited_url <- "https://github.com/USGS-R/drb-network-prep/blob/940073e8d77c911b6fb9dc4e3657aeab1162a158/2_process/out/GFv1_catchments_edited.gpkg?raw=true")

Reply:
Like this idea. I'll incorporate/

Reply:
I've incorporated this. Nice to have the direct url. Does the gpkg get stored in a temp local location?